Use a secure variant of `memset`

howjmay opened the pull request:
#835 
memest_s has been supported since C11, so I implemented memset_safe for earlier C sandard

howjmay committed:
Implement safe memset

Since memset_s is supported after C11. Therefore, I implement this one for the older version.

howjmay committed:
replace memset with memset_safe

jserv commented on utils/memset_safe.h:
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
```
Bump copyright date.

howjmay committed:
correct to the right date

th0br0 commented on utils/memset_safe.c:
```diff
@@ -0,0 +1,26 @@
+    return RC_OOM;
+  }
+
+  volatile unsigned char *ptr = dest;
```
This should default to using memset_s if `__STDC_LIB_EXT1__` is defined

th0br0 commented on cclient/http/http.c:
```diff
@@ -117,7 +117,7 @@ static retcode_t read_data_from_iota_service(int sockfd,
       return RC_OK;
     }
-    memset(buffer, 0, sizeof(buffer));
+    memset_safe(buffer, sizeof(buffer), 0, sizeof(buffer));
```
We don't need a safe memset here

th0br0 commented on consensus/milestone_tracker/milestone_tracker.c:
```diff
@@ -439,7 +439,7 @@ retcode_t iota_milestone_tracker_destroy(milestone_tracker_t* const mt) {
   hash243_queue_free(&mt->candidates);
   rw_lock_handle_destroy(&mt->candidates_lock);
-  memset(mt, 0, sizeof(milestone_tracker_t));
+  memset_safe(mt, sizeof(milestone_tracker_t), 0, sizeof(milestone_tracker_t));
```
No need for safe memset

th0br0 commented on utils/containers/bitset.c:
```diff
@@ -17,7 +17,8 @@ void bitset_reset(bitset_t* const bitset) {
     return;
   }
-  memset(bitset->raw_bits, 0, bitset->size * sizeof(*(bitset->raw_bits)));
+  memset_safe(bitset->raw_bits, bitset->size * sizeof(*(bitset->raw_bits)), 0,
```
No need for safe memset

howjmay committed:
Add wrapper function of memset_s

if __STDC_LIB_EXT1__ was defined, then use memset_s instead of self-implemented memset_sae

howjmay committed:
remove unnecessary memset_safe

rajivshah3 commented on utils/memset_safe.h:
```diff
@@ -0,0 +1,36 @@
+retcode_t memset_safe(void *dest, size_t destsz, int ch, size_t count);
+
+#else
+
```
Does this need to be added?
```suggestion
#define __STDC_WANT_LIB_EXT1__ 1
```

According to https://en.cppreference.com/w/c/string/byte/memset:
> As with all bounds-checked functions, memset_s is only guaranteed to be available if `__STDC_LIB_EXT1__` is defined by the implementation and if the user defines `__STDC_WANT_LIB_EXT1__` to the integer constant 1 before including string.h.

howjmay replied:
thank you for advice

howjmay committed:
define __STDC_WANT_LIB_EXT1__

th0br0 commented on utils/memset_safe.h:
```diff
@@ -0,0 +1,37 @@
+
+#else
+
+#define __STDC_WANT_LIB_EXT1__ 1
```
this is in the wrong location now. 
needs to be at top of the file (i.e. around line 11)

thibault-martinez commented on common/curl-p/curl_p_s.h:
```diff
@@ -9,6 +9,7 @@
 #define __CURL_S_H__
 
 #include "common/trinary/trits.h"
+#include "urils/memset_safe.h"
```
Typo

howjmay replied:
my fault

thibault-martinez commented on utils/memset_safe.c:
```diff
@@ -0,0 +1,30 @@
+    return RC_OOM;
+  } else if (count > destsz) {
+    return RC_OOM;
+  }
```
It doesn't make sense to return `RC_OOM` (Out Of Memory) here.
We shoule return errors that are similar to the real `memset_s`.
```
     [EINVAL]           The s argument was a null pointer.

     [E2BIG]            One or both of smax or n was larger than RSIZE_MAX.

     [EOVERFLOW]        n was larger than smax.
```

howjmay committed:
fix errors

th0br0 committed:
utils: memset_safe cleanup
